File: 05-prototype/05-prototype.php
<?php

interface ComputerPrototype
{
    public function clone(): ComputerPrototype;
    public function showDetails(): void;
}

class GamingPC implements ComputerPrototype
{
    public function __construct(private string $gpu, private int $ram)
    {
    }

    public function clone(): ComputerPrototype
    {
        return new GamingPC($this->gpu, $this->ram);
    }

    public function showDetails(): void
    {
        echo "Gaming PC - GPU: {$this->gpu}, RAM: {$this->ram}GB\n";
    }
}

$gamingPC = new GamingPC("RTX 4090", 32);
$gamingPCclone = $gamingPC->clone();

var_dump($gamingPC, $gamingPCclone);

/* object(GamingPC)#2276 (2) { */
/*   ["gpu":"GamingPC":private]=> */
/*   string(8) "RTX 4090" */
/*   ["ram":"GamingPC":private]=> */
/*   int(32) */
/* } */
/* object(GamingPC)#2275 (2) { */
/*   ["gpu":"GamingPC":private]=> */
/*   string(8) "RTX 4090" */
/*   ["ram":"GamingPC":private]=> */
/*   int(32) */
/* } */

File: 05-prototype/05-prototype.java
interface Computer {
    public String showDetails();
    public Computer clone();
}

class GamingPC implements Computer {
    private String gpu;
    private String ram;

    public GamingPC(String gpu, String ram) {
        this.gpu = gpu;
        this.ram = ram;
    }

    @Override
    public String showDetails() {
        return "GPU: " + this.gpu + "RAM: " + this.ram;
    }

    @Override
    public Computer clone() {
        return new GamingPC(this.gpu, this.ram);
    }
}

public class Main {
    public static void main(String []args) {
        GamingPC gamingPC = new GamingPC("3.4GHz","16GB");
        System.out.println(gamingPC.showDetails());

        Computer gamingPCclone = gamingPC.clone();
        System.out.println(gamingPCclone.showDetails());

        System.out.println(gamingPC);
        System.out.println(gamingPCclone);
    }
}
